politeiad/backendv2/tstorebe/plugins/pi/hooks.go

```diff
@@ -390,11 +395,246 @@ func (p *piPlugin) voteSummary(token []byte) (*ticketvote.SummaryReply, error) {
+	err = json.Unmarshal([]byte(reply), &gar)
+	if err != nil {
+		return nil, err
+	}
+	return &gar, nil
+}
+
+// isAncestorOf returns whether the given parentID is an ancestor of the given
+// childID comment, by traveling the comment tree defined by
+// the ParentID of the child comment.
+func (p *piPlugin) isAncestorOf(ancestorID, childID uint32, cs []comments.Comment) bool {
+	if ancestorID == childID {
+		return true
+	}
+	// Convert comments slice to a map
+	commentsMap := make(map[uint32]comments.Comment, len(cs))
+	for _, c := range cs {
+		commentsMap[c.CommentID] = c
+	}
+	// Travel the comment tree to search for an ancestor
+	for {
+		current := commentsMap[childID]
+		// If we reach the tree haed without visiting the provided
+		// ancestorID then it is not an ancestor.
+		if current.ParentID == 0 {
+			break
+		}
+		// Check if next parent in the tree is the ancestor
+		if current.ParentID == ancestorID {
+			return true
+		}
+		childID = current.ParentID
+	}
+	return false
+}
+
+// latestAuthorUpdate gets the latest author update on a record, if
+// the record has no author update it returns nil.
+func (p *piPlugin) latestAuthorUpdate(token []byte, cs []comments.Comment) (*comments.Comment, error) {
+	var latestAuthorUpdate comments.Comment
+	for _, c := range cs {
+		if c.ExtraDataHint != pi.ProposalUpdateHint {
+			continue
+		}
+		if c.Timestamp > latestAuthorUpdate.Timestamp {
+			latestAuthorUpdate = c
+		}
+	}
+	return &latestAuthorUpdate, nil
+}
+
+// recordAuthor returns the author's userID of the record associated with
+// the provided token.
+func (p *piPlugin) recordAuthor(token []byte) (string, error) {
+	reply, err := p.backend.PluginRead(token, usermd.PluginID,
+		usermd.CmdAuthor, "")
+	if err != nil {
+		return "", err
+	}
+	var ar usermd.AuthorReply
+	err = json.Unmarshal([]byte(reply), &ar)
+	if err != nil {
+		return "", err
+	}
+	return ar.UserID, nil
+}
+
+// commentVoteAllowedOnApprovedProposal verifies that the given comment
+// vote is allowed on the approved proposal associated with the provided
+// token.
+func (p *piPlugin) commentVoteAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var v comments.Vote
+	err := json.Unmarshal([]byte(payload), &v)
+	if err != nil {
+		return err
+	}
+	if !p.isAncestorOf(latestAuthorUpdate.CommentID, v.CommentID,
+		cs) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeVoteStatusInvalid),
+			ErrorContext: "vote has ended; comment votes are allowed only on " +
+				"the latest author update thread",
+		}
+	}
+	return nil
+}
+
+// commentNewAllowedOnApprovedProposal verifies that the given new comment
+// is allowed on the approved proposal associated with the provided
+// token.
+func (p *piPlugin) commentNewAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
```

Comment:
The way you have this function written is not ideal from a performance standpoint.  You typically want to perform all validation that doesn't require network requests before you perform validation that requires network requests.  You should assume that any external calls, like the record author call, will result in a network request.

I would do this more along the lines of:

```
	// A new comment on an approved proposal must either be a update
	// from the author (parent ID will be 0) or a reply to the latest
	// author update.
	isUpdateReply := isAncestorOf(latestAuthorUpdate.CommentID, n.ParentID, cs)
	switch {
	case n.ParentID == 0:
		// This might be an update from the author. Continue below to
		// check.

	case isUpdateReply:
		// This is a reply to the latest update. This is allowed.
		return nil

	case !isUpdateReply:
		// This is a reply, but is not a reply to the latest update. This
		// is not allowed.
		return backend.PluginError{
			PluginID:  pi.PluginID,
			ErrorCode: uint32(pi.ErrorCodeVoteStatusInvalid),
			ErrorContext: "vote has ended; replies are allowed only on " +
				"the latest author update thread",
		}

	default:
		// This should not happen
		return errors.Errorf("unknown comment write state")
	}
```

The user error in code snipped above should probably be changed.  See the comment I made in the `commentVotesAllowedOnApprovedProposal` function.